homeassistant/components/habitica.py

```diff
@@ -0,0 +1,161 @@
+
+
+def has_all_unique_users_names(value):
+    """Validate that all user's names are unique and set if any is set."""
+    names = [user.get(CONF_NAME) for user in value]
+    if None in names and any(name is not None for name in names):
+        raise vol.Invalid(
+            'user names of all users must be set if any is set')
+    if not all(name is None for name in names):
+        has_unique_values(names)
+    return value
+
+
+INSTANCE_LIST_SCHEMA = vol.All(
+    cv.ensure_list,
+    has_all_unique_users,
+    has_all_unique_users_names,
+    [INSTANCE_SCHEMA])
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: INSTANCE_LIST_SCHEMA
+}, extra=vol.ALLOW_EXTRA)
+
+SERVICE_API_CALL = 'api_call'
+ATTR_NAME = CONF_NAME
+ATTR_PATH = CONF_PATH
+ATTR_ARGS = "args"
+EVENT_API_CALL_SUCCESS = "{0}_{1}_{2}".format(
+    DOMAIN, SERVICE_API_CALL, "success")
+
+SERVICE_API_CALL_SCHEMA = vol.Schema({
+    vol.Required(ATTR_NAME): str,
+    vol.Required(ATTR_PATH): vol.All(cv.ensure_list, [str]),
+    vol.Optional(ATTR_ARGS): dict
+})
+
+
+async def async_setup(hass, config):
+    """Set up the habitica service."""
+    conf = config[DOMAIN]
+    data = hass.data[DOMAIN] = {}
+    websession = async_get_clientsession(hass)
+    from habitipy.aio import HabitipyAsync
+
+    class HAHabitipyAsync(HabitipyAsync):
+        """Closure API class to hold session."""
+
+        def __call__(self, **kwargs):
+            return super().__call__(websession, **kwargs)
+
+    for instance in conf:
+        url = instance[CONF_URL]
+        username = instance[CONF_API_USER]
+        password = instance[CONF_API_KEY]
+        name = instance.get(CONF_NAME)
+        config_dict = {"url": url, "login": username, "password": password}
+        api = HAHabitipyAsync(config_dict)
+        user = await api.user.get()
+        if name is None:
+            name = user['profile']['name']
+        if name in data:
+            _LOGGER.warning(
+                "Same user is specified twice: %s", name)
+        data[name] = api
+        if CONF_SENSORS in instance:
+            hass.async_create_task(
+                discovery.async_load_platform(
+                    hass, "sensor", DOMAIN,
+                    {"name": name, "sensors": instance[CONF_SENSORS]},
+                    config))
+
+    async def handle_api_call(call):
+        name = call.data[ATTR_NAME]
+        path = call.data[ATTR_PATH]
+        api = hass.data[DOMAIN].get(name)
+        if api is None:
+            _LOGGER.error(
+                "API_CALL: User '%s' not configured", name)
+            return
+        try:
+            for element in path:
+                api = api[element]
+        except KeyError:
+            _LOGGER.error(
+                "API_CALL: Path %s is invalid"
+                " for api on '{%s}' element", path, element)
+            return
+        kwargs = call.data.get(ATTR_ARGS, {})
+        data = await api(**kwargs)
+        hass.bus.fire(EVENT_API_CALL_SUCCESS, {
```

Comment:
Use the async version of this method, ie `async_fire`, since we're in a coroutine.